File: go/anyonce/key_test.go
package anyonce_test

import (
	"errors"
	"strings"
	"testing"

	"github.com/sns45/anyonce/go/anyonce"
)

func TestValidateKey(t *testing.T) {
	t.Run("REQ-CORE-2: accepts printable ASCII from 1 to 255 bytes", func(t *testing.T) {
		for _, k := range []string{"a", "!~", strings.Repeat("a", anyonce.MaxKeyBytes)} {
			if err := anyonce.ValidateKey(k, false); err != nil {
				t.Fatalf("%q: %v", k, err)
			}
		}
	})
	rejected := map[string]string{
		"empty":                   "",
		"256 bytes":               strings.Repeat("a", 256),
		"space without sf-string": "a b",
		"control char":            "a\x01b",
		"tab":                     "a\tb",
		"DEL":                     "a\x7fb",
		"non-ASCII":               "café",
	}
	for name, key := range rejected {
		t.Run("REQ-CORE-2: rejects "+name, func(t *testing.T) {
			if err := anyonce.ValidateKey(key, false); !errors.Is(err, anyonce.ErrInvalidKey) {
				t.Fatalf("want ErrInvalidKey, got %v", err)
			}
		})
	}
	t.Run("REQ-CORE-2: allows space only when allowSpace is set", func(t *testing.T) {
		if err := anyonce.ValidateKey("a b", true); err != nil {
			t.Fatal(err)
		}
		if err := anyonce.ValidateKey("a b", false); err == nil {
			t.Fatal("expected rejection")
		}
	})
}

func TestParseKey(t *testing.T) {
	cases := []struct {
		name, in string
		syntax   anyonce.Syntax
		want     string
		wantErr  bool
	}{
		{"REQ-CORE-2: lenient accepts a bare token", "abc-123", anyonce.SyntaxLenient, "abc-123", false},
		{"REQ-CORE-2: lenient trims spaces", "  abc  ", anyonce.SyntaxLenient, "abc", false},
		{"REQ-CORE-3: lenient strips quotes and unescapes", `"a\"b"`, anyonce.SyntaxLenient, `a"b`, false},
		{"REQ-CORE-3: lenient keeps a space inside an sf-string", `"with space"`, anyonce.SyntaxLenient, "with space", false},
		{"REQ-CORE-3: strict accepts an sf-string", `"abc"`, anyonce.SyntaxStrict, "abc", false},
		{"REQ-CORE-3: strict rejects a bare token", "abc", anyonce.SyntaxStrict, "", true},
		{"REQ-CORE-3: unterminated quote is invalid in strict", `"abc`, anyonce.SyntaxStrict, "", true},
		{"REQ-CORE-3: unterminated quote is invalid in lenient", `"abc`, anyonce.SyntaxLenient, "", true},
		{"REQ-CORE-2: empty sf-string is invalid", `""`, anyonce.SyntaxLenient, "", true},
		{"REQ-CORE-2: 256 byte bare key is invalid", strings.Repeat("a", 256), anyonce.SyntaxLenient, "", true},
		{"REQ-CORE-2: quoted key over 255 bytes is invalid", `"` + strings.Repeat("a", 256) + `"`, anyonce.SyntaxStrict, "", true},
		{"REQ-CORE-2: surrounding spaces and tabs are optional whitespace and are discarded", "\t abc \t", anyonce.SyntaxLenient, "abc", false},
		{"REQ-CORE-2: other whitespace around the value is not optional whitespace and is rejected", "\x0babc", anyonce.SyntaxLenient, "", true},
		{"REQ-CORE-2: a non-breaking space around a quoted key is rejected", "\xc2\xa0\"abc\"", anyonce.SyntaxStrict, "", true},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got, err := anyonce.ParseKey(tc.in, tc.syntax)
			if tc.wantErr {
				if !errors.Is(err, anyonce.ErrInvalidKey) {
					t.Fatalf("want ErrInvalidKey, got %v", err)
				}
				return
			}
			if err != nil || got != tc.want {
				t.Fatalf("got %q, %v; want %q", got, err, tc.want)
			}
		})
	}
}

File: packages/core/test/key.test.ts
import { describe, expect, test } from 'bun:test';
import { MAX_KEY_BYTES, parseKey, validateKey } from '../src/key';

describe('validateKey', () => {
  test('REQ-CORE-2: accepts printable ASCII from 1 to 255 bytes', () => {
    expect(validateKey('a')).toEqual({ ok: true });
    expect(validateKey('!~')).toEqual({ ok: true });
    expect(validateKey('a'.repeat(MAX_KEY_BYTES))).toEqual({ ok: true });
  });

  const rejected: Array<[string, string, string]> = [
    ['empty', '', 'empty'],
    ['256 bytes', 'a'.repeat(256), 'exceeds 255'],
    ['space without sf-string', 'a b', 'outside printable ASCII'],
    ['control char', 'a\x01b', 'outside printable ASCII'],
    ['tab', 'a\tb', 'outside printable ASCII'],
    ['DEL', 'a\x7fb', 'outside printable ASCII'],
    ['non-ASCII', 'café', 'outside printable ASCII'],
    ['emoji', 'k\u{1F600}', 'outside printable ASCII'],
  ];
  for (const [label, key, reason] of rejected) {
    test(`REQ-CORE-2: rejects ${label}`, () => {
      const result = validateKey(key);
      expect(result.ok).toBe(false);
      if (!result.ok) expect(result.reason).toMatch(reason);
    });
  }

  test('REQ-CORE-2: allows space only when allowSpace is set', () => {
    expect(validateKey('a b', true)).toEqual({ ok: true });
    expect(validateKey(' ', true)).toEqual({ ok: true });
    expect(validateKey('a b', false).ok).toBe(false);
  });
});

describe('parseKey', () => {
  test('REQ-CORE-2: lenient accepts a bare token', () => {
    expect(parseKey('abc-123', 'lenient')).toEqual({ ok: true, key: 'abc-123' });
    expect(parseKey('  abc  ', 'lenient')).toEqual({ ok: true, key: 'abc' });
  });

  test('REQ-CORE-3: lenient strips quotes and unescapes an sf-string', () => {
    expect(parseKey('"abc"', 'lenient')).toEqual({ ok: true, key: 'abc' });
    expect(parseKey('"a\\"b"', 'lenient')).toEqual({ ok: true, key: 'a"b' });
    expect(parseKey('"with space"', 'lenient')).toEqual({ ok: true, key: 'with space' });
  });

  test('REQ-CORE-3: strict requires an sf-string', () => {
    expect(parseKey('"abc"', 'strict')).toEqual({ ok: true, key: 'abc' });
    const bare = parseKey('abc', 'strict');
    expect(bare.ok).toBe(false);
    if (!bare.ok) expect(bare.code).toBe('invalid-key');
  });

  test('REQ-CORE-3: an unterminated quote is invalid in both modes', () => {
    expect(parseKey('"abc', 'strict').ok).toBe(false);
    expect(parseKey('"abc', 'lenient').ok).toBe(false);
  });

  test('REQ-CORE-2: an empty sf-string is invalid-key', () => {
    const result = parseKey('""', 'lenient');
    expect(result.ok).toBe(false);
    if (!result.ok) expect(result.reason).toMatch('empty');
  });

  test('REQ-CORE-2: a 256 byte bare key is invalid-key', () => {
    const result = parseKey('a'.repeat(256), 'lenient');
    expect(result.ok).toBe(false);
    if (!result.ok) expect(result.code).toBe('invalid-key');
  });

  test('REQ-CORE-2: a quoted key over 255 bytes is invalid-key', () => {
    expect(parseKey(`"${'a'.repeat(256)}"`, 'strict').ok).toBe(false);
  });

  test('REQ-CORE-2: surrounding spaces and tabs are optional whitespace and are discarded', () => {
    expect(parseKey('\t abc \t', 'lenient')).toEqual({ ok: true, key: 'abc' });
  });

  test('REQ-CORE-2: other whitespace around the value is not optional whitespace and is rejected', () => {
    expect(parseKey('\x0babc', 'lenient').ok).toBe(false);
    expect(parseKey('\xa0"abc"', 'strict').ok).toBe(false);
  });
});
